File: app/src/main/java/com/alphadevelopmentsolutions/frcscout/Fragments/QuickStatsFragment.java
package com.alphadevelopmentsolutions.frcscout.Fragments;

import android.app.Fragment;
import android.content.Context;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TableLayout;
import android.widget.TableRow;
import android.widget.TextView;

import com.alphadevelopmentsolutions.frcscout.Classes.Team;
import com.alphadevelopmentsolutions.frcscout.Interfaces.StatsKeys;
import com.alphadevelopmentsolutions.frcscout.R;

import java.util.ArrayList;
import java.util.HashMap;


/**
 * A simple {@link Fragment} subclass.
 * Activities that contain this fragment must implement the
 * {@link QuickStatsFragment.OnFragmentInteractionListener} interface
 * to handle interaction events.
 * Use the {@link QuickStatsFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
public class QuickStatsFragment extends MasterFragment
{
    private OnFragmentInteractionListener mListener;

    public QuickStatsFragment()
    {
        // Required empty public constructor
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @param team
     * @return A new instance of fragment QuickStatsFragment.
     */
    // TODO: Rename and change types and number of parameters
    public static QuickStatsFragment newInstance(@NonNull Team team)
    {
        QuickStatsFragment fragment = new QuickStatsFragment();
        Bundle args = new Bundle();
        args.putString(ARG_TEAM_JSON, toJson(team));
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);

        loadStatsThread = new Thread(new Runnable()
        {
            @Override
            public void run()
            {

                joinLoadingThread();

                //get the stats from the team object
                HashMap<String, HashMap<String, Double>> allStats = team.getStats(database, event);
                minTableRows = new ArrayList<>();
                avgTableRows = new ArrayList<>();
                maxTableRows = new ArrayList<>();

                //generate padding based on screen size
                int padding = context.getResources().getDimensionPixelSize(R.dimen.material_padding);

                //iterate through each category in the stats interface
                for(final String CATEGORY_KEY : StatsKeys.STATS_CATEGORY_ORDER_KEYS)
                {
                    //get the stats for each category
                    HashMap<String, Double> stats = allStats.get(CATEGORY_KEY);

                    //iterate through each stat in the stats interface
                    for(String KEY : StatsKeys.STATS_ORDER_KEYS)
                    {
                        //get the stat
                        double stat = stats.get(KEY);

                        //generate the textview
                        TextView textView = new TextView(context);
                        textView.setText(String.valueOf(((!CATEGORY_KEY.equals(StatsKeys.AVG)) ? (int) Math.ceil(stat) : Math.round(stat * 100.00) / 100.00)));
                        textView.setPadding(padding, padding, padding, padding);

                        //create the tablerow to add to the table layout
                        final TableRow tableRow = new TableRow(context);

                        //color logic
                        if(KEY.equals(StatsKeys.AUTO_EXIT_HAB))
                        {
                            if(stat > 0)
                                tableRow.setBackgroundColor(context.getColor(R.color.good));
                            else
                                tableRow.setBackgroundColor(context.getColor(R.color.bad));
                        }

                        else if(KEY.equals(StatsKeys.AUTO_HATCH))
                        {
                            if(stat > 0)
                                tableRow.setBackgroundColor(context.getColor(R.color.good));
                            else
                                tableRow.setBackgroundColor(context.getColor(R.color.bad));
                        }

                        else if(KEY.equals(StatsKeys.AUTO_HATCH_DROPPED))
                        {
                            if(stat > 0)
                                tableRow.setBackgroundColor(context.getColor(R.color.bad));
                            else
                                tableRow.setBackgroundColor(context.getColor(R.color.good));
                        }

                        else if(KEY.equals(StatsKeys.AUTO_CARGO))
                        {
                            if(stat > 0)
                                tableRow.setBackgroundColor(context.getColor(R.color.good));
                            else
                                tableRow.setBackgroundColor(context.getColor(R.color.bad));
                        }

                        else if(KEY.equals(StatsKeys.AUTO_CARGO_DROPPED))
                        {
                            if(stat > 0)
                                tableRow.setBackgroundColor(context.getColor(R.color.bad));
                            else
                                tableRow.setBackgroundColor(context.getColor(R.color.good));
                        }

                        else if(KEY.equals(StatsKeys.TELEOP_HATCH))
                        {
                            if(stat > 0)
                                tableRow.setBackgroundColor(context.getColor(R.color.good));
                            else
                                tableRow.setBackgroundColor(context.getColor(R.color.bad));
                        }

                        else if(KEY.equals(StatsKeys.TELEOP_HATCH_DROPPED))
                        {
                            if(stat > 0)
                                tableRow.setBackgroundColor(context.getColor(R.color.bad));
                            else
                                tableRow.setBackgroundColor(context.getColor(R.color.good));
                        }

                        else if(KEY.equals(StatsKeys.TELEOP_CARGO))
                        {
                            if(stat > 0)
                                tableRow.setBackgroundColor(context.getColor(R.color.good));
                            else
                                tableRow.setBackgroundColor(context.getColor(R.color.bad));
                        }

                        else if(KEY.equals(StatsKeys.TELEOP_CARGO_DROPPED))
                        {
                            if(stat > 0)
                                tableRow.setBackgroundColor(context.getColor(R.color.bad));
                            else
                                tableRow.setBackgroundColor(context.getColor(R.color.good));
                        }

                        else if(KEY.equals(StatsKeys.RETURNED_HAB))
                        {
                            if(stat > 0)
                                tableRow.setBackgroundColor(context.getColor(R.color.good));
                            else
                                tableRow.setBackgroundColor(context.getColor(R.color.bad));
                        }

                        else if(KEY.equals(StatsKeys.RETURNED_HAB_FAILS))
                        {
                            if(stat > 0)
                                tableRow.setBackgroundColor(context.getColor(R.color.bad));
                            else
                                tableRow.setBackgroundColor(context.getColor(R.color.good));
                        }

                        else if(KEY.equals(StatsKeys.DEFENSE_RATING))
                        {
                            if(stat >= 0 && stat <= 1.6)
                                tableRow.setBackgroundColor(context.getColor(R.color.bad));
                            else if (stat <= 3.2)
                                tableRow.setBackgroundColor(context.getColor(R.color.warning));
                            else
                                tableRow.setBackgroundColor(context.getColor(R.color.good));
                        }

                        else if(KEY.equals(StatsKeys.OFFENSE_RATING))
                        {
                            if(stat >= 0 && stat <= 1.6)
                                tableRow.setBackgroundColor(context.getColor(R.color.bad));
                            else if (stat <= 3.2)
                                tableRow.setBackgroundColor(context.getColor(R.color.warning));
                            else
                                tableRow.setBackgroundColor(context.getColor(R.color.good));
                        }

                        else if(KEY.equals(StatsKeys.DRIVE_RATING))
                        {
                            if(stat >= 0 && stat <= 1.6)
                                tableRow.setBackgroundColor(context.getColor(R.color.bad));
                            else if (stat <= 3.2)
                                tableRow.setBackgroundColor(context.getColor(R.color.warning));
                            else
                                tableRow.setBackgroundColor(context.getColor(R.color.good));
                        }

                        //set the gravity to center the text and add it
                        tableRow.setGravity(Gravity.CENTER);
                        tableRow.addView(textView);

                        //add each tablerow to the array
                        if(CATEGORY_KEY.equals(StatsKeys.MIN))
                            minTableRows.add(tableRow);
                        else if(CATEGORY_KEY.equals(StatsKeys.AVG))
                            avgTableRows.add(tableRow);
                        else
                            maxTableRows.add(tableRow);
                    }

                }

            }

        });

        loadStatsThread.start();
    }

    private TableLayout statsMinTableLayout;
    private TableLayout statsMaxTableLayout;
    private TableLayout statsAvgTableLayout;

    private Thread loadStatsThread;

    private ArrayList<TableRow> minTableRows;
    private ArrayList<TableRow> avgTableRows;
    private ArrayList<TableRow> maxTableRows;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState)
    {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_quick_stats, container, false);

        statsMinTableLayout = view.findViewById(R.id.StatsMinTableLayout);
        statsMaxTableLayout = view.findViewById(R.id.StatsMaxTableLayout);
        statsAvgTableLayout = view.findViewById(R.id.StatsAvgTableLayout);

        new Thread(new Runnable()
        {
            @Override
            public void run()
            {
                try
                {
                    //make sure the load stats thread has finish before attempting
                    loadStatsThread.join();

                    //check if the table rows have a parent, if they do remove every view
                    //this includes table rows from min, max and avg
                    //mainly has to be done when the view is created on a rotate or backstack pop
                    if(minTableRows.get(0).getParent() != null)
                        ((ViewGroup) minTableRows.get(0).getParent()).removeAllViews();

                    if(avgTableRows.get(0).getParent() != null)
                        ((ViewGroup) avgTableRows.get(0).getParent()).removeAllViews();

                    if(maxTableRows.get(0).getParent() != null)
                        ((ViewGroup) maxTableRows.get(0).getParent()).removeAllViews();

                    //iterate through each of the table rows
                    for(int i = 0; i < minTableRows.size(); i++)
                    {

                        //add each category row
                        final int finalI = i;
                        context.runOnUiThread(new Runnable()
                        {
                            @Override
                            public void run()
                            {
                                //add each view to the respective tablelayout
                                statsMinTableLayout.addView(minTableRows.get(finalI));
                                statsAvgTableLayout.addView(avgTableRows.get(finalI));
                                statsMaxTableLayout.addView(maxTableRows.get(finalI));

                            }
                        });
                    }
                } catch (InterruptedException e)
                {
                    e.printStackTrace();
                }

            }
        }).start();

        return view;
    }

    // TODO: Rename method, update argument and hook method into UI event
    public void onButtonPressed(Uri uri)
    {
        if (mListener != null)
        {
            mListener.onFragmentInteraction(uri);
        }
    }

    @Override
    public void onAttach(Context context)
    {
        super.onAttach(context);
        if (context instanceof OnFragmentInteractionListener)
        {
            mListener = (OnFragmentInteractionListener) context;
        } else
        {
            throw new RuntimeException(context.toString()
                    + " must implement OnFragmentInteractionListener");
        }
    }

    @Override
    public void onDetach()
    {
        super.onDetach();
        mListener = null;
    }

    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     * <p>
     * See the Android Training lesson <a href=
     * "http://developer.android.com/training/basics/fragments/communicating.html"
     * >Communicating with Other Fragments</a> for more information.
     */
    public interface OnFragmentInteractionListener
    {
        // TODO: Update argument type and name
        void onFragmentInteraction(Uri uri);
    }
}

File: app/src/main/java/com/alphadevelopmentsolutions/frcscout/Fragments/QuickStatsFragment.kt
package com.alphadevelopmentsolutions.frcscout.Fragments

import android.graphics.Color
import android.graphics.Typeface
import android.os.Bundle
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import com.alphadevelopmentsolutions.frcscout.Classes.Database
import com.alphadevelopmentsolutions.frcscout.Classes.Tables.Match
import com.alphadevelopmentsolutions.frcscout.Classes.Tables.ScoutCardInfo
import com.alphadevelopmentsolutions.frcscout.Classes.Tables.ScoutCardInfoKey
import com.alphadevelopmentsolutions.frcscout.Classes.Tables.Team
import com.alphadevelopmentsolutions.frcscout.R
import com.github.mikephil.charting.charts.LineChart
import com.github.mikephil.charting.components.Legend
import com.github.mikephil.charting.components.XAxis
import com.github.mikephil.charting.data.Entry
import com.github.mikephil.charting.data.LineData
import com.github.mikephil.charting.data.LineDataSet
import com.github.mikephil.charting.formatter.IFillFormatter
import com.github.mikephil.charting.formatter.ValueFormatter
import com.github.mikephil.charting.interfaces.datasets.ILineDataSet
import kotlinx.android.synthetic.main.fragment_quick_stats.view.*
import java.util.*
import java.util.concurrent.CountDownLatch
import kotlin.collections.ArrayList
import kotlin.collections.HashMap
import kotlin.collections.LinkedHashMap


class QuickStatsFragment : MasterFragment()
{
    override fun onBackPressed(): Boolean
    {
        return false
    }

    private lateinit var loadStatsThread: Thread

    private lateinit var tempStatChart: LineChart

    private val statCharts = ArrayList<LinearLayout>()

    override fun onCreate(savedInstanceState: Bundle?)
    {
        super.onCreate(savedInstanceState)

        loadStatsThread = Thread(Runnable {
            loadingThread.join()


            val matches = Match.getObjects(event!!, null, team!!, database, Database.SortDirection.ASC) //matches for this team
            val scoutCardInfoKeys = ScoutCardInfoKey.getObjects(year!!, null, database) //scout card info keys for this year
            val scoutCardInfos = ScoutCardInfo.getObjects(event, null, null, null, null, false, database) //scout card infos for this event

            //gets the height in DP to display
            val dp = fun(height: Int): Int
            {
                return (height * context.resources.displayMetrics.density + 0.5f).toInt()
            }

            if (!scoutCardInfoKeys.isNullOrEmpty() && !matches.isNullOrEmpty() && !scoutCardInfos.isNullOrEmpty())
            {
                val eventStatsHashMap = event!!.getStats(year, scoutCardInfoKeys, scoutCardInfos, database) //get event avg data

                //get match stats data for each match
                val matchStatsHashMap = HashMap<String, HashMap<String, Double>>()
                for(match in matches)
                    matchStatsHashMap[match.toString()] = match.getStats(year, event, scoutCardInfoKeys, scoutCardInfos, database)

                val teamStatsHashMap = team!!.getStats(year, event, matches, scoutCardInfoKeys, scoutCardInfos, database) // get team stats data
                
                // get all the states from the scout card info keys
                val scoutInfoKeyStates = LinkedHashMap<String, ArrayList<ScoutCardInfoKey>>().apply {
                    
                    for(scoutCardInfoKey in scoutCardInfoKeys)
                    {
                        if(scoutCardInfoKey.includeInStats)
                            this[scoutCardInfoKey.keyState] = (this[scoutCardInfoKey.keyState] ?: ArrayList()).apply {
                                add(scoutCardInfoKey)
                            }
                    }
                }
                
                //iterate through all the info key states
                for(scoutCardInfoKeyState in scoutInfoKeyStates.values)
                {
                    //latch used to hold the thread until the runOnUiThread has finished
                    val latch = CountDownLatch(1)

                    context.runOnUiThread {

                        tempStatChart = LineChart(context)

                        latch.countDown()

                    }

                    latch.await()

                    val statChart = tempStatChart

                    //set stat chart properties
                    with(statChart)
                    {
                        layoutParams = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
                        layoutParams.height = dp(250)
                        setTouchEnabled(false)
                        setDrawGridBackground(true)
                        isDragEnabled = false
                        setPinchZoom(false)
                        isDoubleTapToZoomEnabled = false
                        description.isEnabled = false
                        xAxis.position = XAxis.XAxisPosition.BOTTOM
                        xAxis.granularity = 1f
                        xAxis.labelRotationAngle = -45f
                        xAxis.labelCount = matches.size
                        xAxis.valueFormatter = object: ValueFormatter(){

                            override fun getFormattedValue(value: Float): String
                            {
                                return matches[value.toInt()].toString()
                            }
                        }
                        axisLeft.granularity = 1f
                        axisRight.isEnabled = false
                        legend.verticalAlignment = Legend.LegendVerticalAlignment.TOP
                        legend.horizontalAlignment = Legend.LegendHorizontalAlignment.CENTER
                        legend.yOffset = 15f
                    }
                    
                    //list of entries to show on the stat chart
                    val teamStats = ArrayList<Entry>()
                    val matchStats = ArrayList<Entry>()
                    val eventAvg = ArrayList<Entry>()
                    
                    //array to hold entries
                    val dataSets = ArrayList<ILineDataSet>()
                    
                    //data to hold data sets
                    var data: LineData
                    val filler = IFillFormatter { _, _ -> statChart.axisLeft.axisMinimum }
                    
                    //sets default data set properties
                    val formatDataSets = fun (dataSet: LineDataSet)
                    {
                        with(dataSet)
                        {
                            mode = LineDataSet.Mode.HORIZONTAL_BEZIER
                            setDrawValues(false)
                            lineWidth = 2f
                            circleRadius = 3f
                            setDrawCircleHole(false)
                            valueTextSize = 9f
                            setDrawFilled(true)
                            fillFormatter = filler
                        }
                        
                    }

                    val teamDataSet = LineDataSet(ArrayList<Entry>(), getString(R.string.team_average)).apply{
                        formatDataSets(this)
                        color = Color.parseColor("#03A9F4")
                        setCircleColor(Color.parseColor("#03A9F4"))
                        fillColor = Color.argb(49, 3, 169, 244)
                    }
                    

                    val matchDataSet = LineDataSet(ArrayList<Entry>(), getString(R.string.match_average)).apply{
                        formatDataSets(this)
                        color = Color.parseColor("#FF0000")
                        setCircleColor(Color.parseColor("#FF0000"))
                        fillColor = Color.argb(49, 244, 0, 0)
                    }
                    

                    val eventAvgDataSet = LineDataSet(ArrayList<Entry>(), getString(R.string.event_average)).apply{
                        formatDataSets(this)
                        color = Color.parseColor("#005685")
                        setDrawCircles(false)
                        setDrawCircleHole(false)
                        setDrawFilled(false)
                        enableDashedLine(25f, 20f, 1f)
                    }
                    

                    //add stat chart to the statcharts arraylist
                    statCharts.add(
                            //add the linear layout to the stat chart
                            LinearLayout(context).apply{
                                orientation = LinearLayout.VERTICAL
                                layoutParams = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
                                setPadding(0, 0, 0, dp(8))

                                //add the textview with the title to the linear layout
                                addView(
                                        TextView(context).apply {
                                            text = scoutCardInfoKeyState[0].keyState
                                            setTypeface(null, Typeface.BOLD)
                                            gravity = Gravity.CENTER
                                            setTextColor(Color.BLACK)
                                        })

                                //add the spinner selector to the linear layout
                                addView(
                                        Spinner(context).apply {

                                            val spinnerArray = ArrayList<String>().apply {

                                                for(scoutCardInfo in scoutInfoKeyStates[scoutCardInfoKeyState[0].keyState]!!)
                                                    add(scoutCardInfo.keyName)
                                            }

                                            adapter = ArrayAdapter<String>(context, android.R.layout.simple_spinner_dropdown_item, spinnerArray)

                                            onItemSelectedListener = object: AdapterView.OnItemSelectedListener{
                                                override fun onNothingSelected(p0: AdapterView<*>?)
                                                {

                                                }

                                                //when new item selected, update the chart
                                                override fun onItemSelected(p0: AdapterView<*>?, p1: View?, p2: Int, p3: Long)
                                                {
                                                    //used to find min / max values
                                                    val chartPoints = ArrayList<Float>()

                                                    teamStats.clear()
                                                    matchDataSet.clear()
                                                    eventAvgDataSet.clear()

                                                    for (i in 0 until matches.size)
                                                    {
                                                        val teamStat = teamStatsHashMap[matches[i].toString()]!![scoutCardInfoKeyState[p2].toString()]!!.toFloat()
                                                        val matchStat = matchStatsHashMap[matches[i].toString()]!![scoutCardInfoKeyState[p2].toString()]!!.toDouble().toFloat()

                                                        chartPoints.add(teamStat)
                                                        chartPoints.add(matchStat)

                                                        teamStats.add(Entry(i.toFloat(), teamStat))
                                                        matchStats.add(Entry(i.toFloat(), matchStat))

                                                        eventAvg.add(Entry(i.toFloat(), eventStatsHashMap[scoutCardInfoKeyState[p2].toString()]!!.toFloat()))
                                                    }

                                                    teamDataSet.values = teamStats
                                                    matchDataSet.values = matchStats
                                                    eventAvgDataSet.values = eventAvg

                                                    dataSets.clear()
                                                    dataSets.add(teamDataSet)
                                                    dataSets.add(matchDataSet)
                                                    dataSets.add(eventAvgDataSet)

                                                    data = LineData(dataSets)

                                                    statChart.data = data
                                                    statChart.axisLeft.axisMinimum = Collections.min(chartPoints)
                                                    statChart.invalidate()
                                                }
                                            }
                                        }
                                )

                                addView(statChart)
                    })
                }
            }
        })

        loadStatsThread.start()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View?
    {
        // Inflate the layout for this fragment
        val view = inflater.inflate(R.layout.fragment_quick_stats, container, false)

        Thread(Runnable {
            loadStatsThread.join()

            context.runOnUiThread{
                for(statChart in statCharts)
                {
                    val chartParent = if(statChart.parent != null) statChart.parent as LinearLayout else null
                    if(chartParent != null && chartParent.childCount > 0)
                        chartParent.removeAllViews()

                    view.StatChartsLinearLayout.addView(statChart)
                }
            }
        }).start()

        return view
    }

    companion object
    {

        /**
         * Creates a new instance
         * @param team to show stats for
         * @return A new instance of fragment [QuickStatsFragment].
         */
        fun newInstance(team: Team): QuickStatsFragment
        {
            val fragment = QuickStatsFragment()
            val args = Bundle()
            args.putString(ARG_TEAM_JSON, toJson(team))
            fragment.arguments = args
            return fragment
        }
    }
}
